Add support to update categories of a message, delta synching, sending of drafts

marlemion opened the pull request:
For some reasons, the setattr method does not work for the categories of a message. Hence, the update_attr_value method does also not work.

Therefore, I have added an explicit method set_categories, which sets item.categories directly.

Commit:
Add support to updat categories of a message

Commit:
Re-enable delta sync of messages in folder

Commit:
(Re-)enable sending of drafts

mortymacs commented on grapi/backend/kopano/message.py:
```diff
@@ -119,6 +129,7 @@ class MessageResource(ItemResource):
         'from': lambda item, arg: set_user_email(item, 'from_', arg),
         'sender': lambda item, arg: set_user_email(item, 'sender', arg),
         'isRead': lambda item, value: update_attr_value(item, "read", value),
+        'categories': lambda item, value: set_categories(item, value),
```
```suggestion
        'categories': lambda item, value: update_attr_value(item, "categories", value),
```
Please use `update_attr_value` instead.

marlemion replied:
Should be fixed in https://github.com/Kopano-dev/grapi/pull/4/commits/ad62849ee1aa29192de26dc0751fa7959b05cccf

mortymacs commented on grapi/backend/kopano/message.py:
```diff
@@ -56,6 +56,16 @@ def update_attr_value(item, attr_name, value):
     setattr(item, attr_name, value)
 
 
+def set_categories(item, value):
```
Please use `update_attr_value` to set or update attributes. So, this function is no longer needed.

marlemion replied:
Should be fixed in https://github.com/Kopano-dev/grapi/pull/4/commits/ad62849ee1aa29192de26dc0751fa7959b05cccf

mortymacs commented on grapi/backend/kopano/message.py:
```diff
@@ -164,11 +175,15 @@ def on_get_item(self, req, resp, folderid=None, itemid=None):
-        store = req.context.server_store[1]
-        item = _item(store, itemid)
-        self.respond(req, resp, item)
+        elif itemid == 'delta':  # TODO move to MailFolder resource somehow?
```
for the `delta` endpoint, please move them in a separate method and then add their endpoints to the route list. So, it will be the `on_get_delta` method within this class and endpoints:
```
GET /me/mailFolders/{id}/messages/delta
GET /users/{id}/mailFolders/{id}/messages/delta
```
source: https://docs.microsoft.com/en-us/graph/api/message-delta?view=graph-rest-1.0&tabs=http#http-request

marlemion replied:
Should be fixed in https://github.com/Kopano-dev/grapi/pull/4/commits/85a92a97405808ed00a36f4620d378b80b234acf

mortymacs reviewed (changes requested):
Hi,
Thank you so much for the contribution!
I've put some comments in your PR.
Cheers,

Commit:
Re-enable members functionm of groups

Commit:
Rework categories

Commit:
Rework send

Commit:
Rework folder delta

marlemion commented:
Hi,

> Thank you so much for the contribution!

Your welcome. Thanks for the nice API work. :)

> I've put some comments in your PR.

I have updated the PR. Could you please go through it again? Also, I had added a small patch https://github.com/Kopano-dev/grapi/pull/4/commits/ce4b292609d638d15e8b0ff1b27b15738c8282ca after I recognized your comments. That would need reworking, too.

Thanks!

mortymacs commented:
> Hi,
> 
> > Thank you so much for the contribution!
> 
> Your welcome. Thanks for the nice API work. :)
> 
> > I've put some comments in your PR.
> 
> I have updated the PR. Could you please go through it again? Also, I had added a small patch [ce4b292](https://github.com/Kopano-dev/grapi/commit/ce4b292609d638d15e8b0ff1b27b15738c8282ca) after I recognized your comments. That would need reworking, too.
> 
> Thanks!

Thanks for your quick update!
I just put a new comment in your new update.

mortymacs commented on grapi/backend/kopano/group.py:
```diff
@@ -51,6 +51,14 @@ def on_get_member_of(self, req, resp):
+    def on_get_members(self, req, resp, groupid):
+        server, _, userid = req.context.server_store
+
+        if not groupid:
```
How it would be possible to have an empty group ID when Falcon validates the URL and makes sure it matches the endpoint?

marlemion replied:
Correct, I have added the check after some testing was unsuccessful. But another test showed that groupid gets correctly handled by falcon (my first, well second, but first real case using falcon - still learning).

Should be fixed in https://github.com/Kopano-dev/grapi/pull/4/commits/10fe048867697fdde0861c029a90e4ff1f1bfae0

Commit:
Remove unnecessary groupid check from on_get_members

marlemion commented:
> Thanks for your quick update!
> I just put a new comment in your new update.

Thank you! I have updated the PR accordingly.

mortymacs commented on grapi/backend/kopano/group.py:
```diff
@@ -51,6 +51,11 @@ def on_get_member_of(self, req, resp):
         self.respond(req, resp, data, self.fields)
 
+    def on_get_members(self, req, resp, groupid):
+        server, _, _ = req.context.server_store
```
```suggestion
        server = req.context.server_store[0]
```

mortymacs commented on grapi/backend/kopano/group.py:
```diff
@@ -51,6 +51,11 @@ def on_get_member_of(self, req, resp):
         data = (user.groups(), DEFAULT_TOP, 0, 0)
         self.respond(req, resp, data, self.fields)
 
+    def on_get_members(self, req, resp, groupid):
```
```suggestion
    def on_get_members(self, req, resp, groupid):
        """Return members by group ID.

        Args:
            req (Request): Falcon request object.
            resp (Response): Falcon response object.
            groupid (str): group ID.
        """
```

mortymacs commented on grapi/backend/kopano/message.py:
```diff
@@ -236,10 +257,26 @@ def on_post_createReplyAll(self, req, resp, folderid=None, itemid=None):
+            resp (Response): Falcon response object.
+            itemid (str): message ID. Defaults to None. itemid value is mandatory.
+        """
+
```
```suggestion
```

mortymacs commented on grapi/backend/kopano/message.py:
```diff
@@ -149,6 +150,25 @@ def on_get_messages(self, req, resp):
+
+        if folderid is None:
+            raise HTTPNotFound()
+
```
```suggestion
```

mortymacs commented on grapi/backend/kopano/message.py:
```diff
@@ -149,6 +150,25 @@ def on_get_messages(self, req, resp):
+        Raises:
+            HTTPNotFound: when folderid is None.
+        """
+
```
```suggestion
```

mortymacs reviewed (changes requested):
Thank you!
I just put a few changes that need to be applied. Please review them then I'll approve the PR.

marlemion committed:
Update grapi/backend/kopano/group.py

Co-authored-by: Morteza NourelahiAlamdari <m@0t1.me>

marlemion committed:
Update grapi/backend/kopano/message.py

Co-authored-by: Morteza NourelahiAlamdari <m@0t1.me>

marlemion committed:
Update 2 grapi/backend/kopano/group.py

Co-authored-by: Morteza NourelahiAlamdari <m@0t1.me>

marlemion committed:
Update grapi/backend/kopano/message.py

Co-authored-by: Morteza NourelahiAlamdari <m@0t1.me>

marlemion committed:
Update 2 grapi/backend/kopano/message.py

Co-authored-by: Morteza NourelahiAlamdari <m@0t1.me>

marlemion commented:
Hi, thanks again. I have committed all suggested changes.

mortymacs commented:
> Hi, thanks again. I have committed all suggested changes.

Thank you!

fbartels commented:
For internal reference https://stash.kopano.io/projects/KC/repos/grapi/pull-requests/171/overview

(has the merge conflict already resolved)

longsleep commented:
> For internal reference https://stash.kopano.io/projects/KC/repos/grapi/pull-requests/171/overview

Merged - thank you!
